_posts/2017-05-30-database-encryption.markdown

```diff
@@ -0,0 +1,172 @@
+     -----END RSA PRIVATE KEY-----
+
+  # List Symmetric Key files in the order of current / latest first
+  ciphers:
+     -
+        # Filename containing Symmetric Encryption Key encrypted using the
+        # RSA public key derived from the private key above
+        key_filename: /etc/rails/.rails.key
+        iv_filename:  /etc/rails/.rails.iv
+
+        # Encryption cipher_name
+        #   Recommended values:
+        #      aes-256-cbc
+        #         256 AES CBC Algorithm. Very strong
+        #         Ruby 1.8.7 MRI Approximately 100,000 encryptions or decryptions per second
+        #         JRuby 1.6.7 with Ruby 1.8.7 Approximately 22,000 encryptions or decryptions per second
+        #      aes-128-cbc
+        #         128 AES CBC Algorithm. Less strong.
+        #         Ruby 1.8.7 MRI Approximately 100,000 encryptions or decryptions per second
+        #         JRuby 1.6.7 with Ruby 1.8.7 Approximately 22,000 encryptions or decryptions per second
+        cipher_name:  aes-256-cbc
+
+```
+
+### On Heroku
+Since we don't have access to file system in Heroku outside of source code, `symmetric-encryption` suggests to set encryption key as an environment variable. Configuration is same as above except we replace `key_filename` with `encrypted_key: "<%= ENV['PRODUCTION_ENCRYPTION_KEY1'] %>"`.
+
+You can use following rake task to generate Heroku specific configuration file:
+```
+rails g symmetric_encryption:heroku_config
+```
+This will create `config/symmetric-encryption.yml` file and also output commands you can run to set _encrypted_ encryption key on Heroku as an environment variable.
+
+
+For someone in order to be able to decode the data:
+
+
+- They have to be able to access our database.
+- They need to be able to access our source code.
+- They need to be able to access key files on disk or Heroku's configuration.
+
+### ActiveRecord Integration
+Symmetric Encryption provides a seamless integration with `ActiveRecord`, we can use `attr_encrypted` helper method to define _encrypted_ fields. Lets say we wanted to encrypt a `Note` model that has `note` and `subject`, you can add following to your `ActiveRecord` model:
+
+```ruby
+# app/models/note.rb
+attr_encrypted :note
+attr_encrypted :subject
+```
+
+This means whenever you set `note` for this model `symmetric-encryption` will set `encrypted_note` field in the database. Whenever you retrieve an instance of this model, `symmetric-encryption` will decrypt `encrypted_note` field and you can access _decrypted_ value by just accessing `note`.
+
+In our case we couldn't use this helper immediately, the issue is, using `encrypted_attr` means we will no longer be able to use un-encrypted fields. In our migration process until 4th step we still use non-encrypted fields and we need to populate them. In order to do that we started with adding a `before_validation` callback to our model to set encrypted fields based on un-encrypted ones.
```

Comment:
...immediately. Using `encrypted_attr` would prevent us from directly accessing the existing, unencrypted fields in our database (which is necessary through step 3 in our approach). To work around this, we started by adding...